File: services/purchase-service/models/po.go
package models

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

type POStatus string

const (
	POStatusDraft     POStatus = "DRAFT"
	POStatusSent      POStatus = "SENT"
	POStatusCompleted POStatus = "COMPLETED"
)

type PurchaseOrder struct {
	gorm.Model
	ID             uint     `gorm:"primaryKey"`
	PONumber       string   `gorm:"unique;not null"`
	PRID           uint     `gorm:"not null;index"`
	VendorID       uint     `gorm:"not null;index"`
	Status         POStatus `gorm:"type:varchar(20);default:'DRAFT'"`
	CreditDay      int      `gorm:"default:0"`
	DueDate        time.Time
	Items          []POItem        `gorm:"foreignKey:POID"`
	VendorSnapshot *VendorSnapshot `gorm:"foreignKey:POID"`
	IsDeleted      bool            `gorm:"default:false"`
	CreatedAt      time.Time
	UpdatedAt      time.Time
	DeletedAt      gorm.DeletedAt `gorm:"index"`
}

type POItem struct {
	gorm.Model
	ID           uint   `gorm:"primaryKey"`
	POID         uint   `gorm:"not null;index"`
	ItemName     string `gorm:"not null"`
	Description  string
	Quantity     int     `gorm:"not null"`
	Unit         string  `gorm:"not null"`
	PricePerUnit float64 `gorm:"type:decimal(15,2);not null"`
	Discount     float64 `gorm:"type:decimal(15,2)"`
	DiscountUnit string  `gorm:"type:varchar(10)"` // % or BAHT
	TotalPrice   float64 `gorm:"type:decimal(15,2);not null"`
	RequiredDate time.Time
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

// GoodsReceived records when items are received for data tracking
type GoodsReceived struct {
	gorm.Model
	ID           uint            `gorm:"primaryKey"`
	POID         uint            `gorm:"not null;index"`
	ReceivedData json.RawMessage `gorm:"type:jsonb"`
	ReceivedAt   time.Time
	CreatedAt    time.Time
}

File: client/src/lib/api/po.ts
import { request } from "./client";

export interface POItem {
    ID: number;
    POID: number;
    CreatedAt: string;
    UpdatedAt: string;
    SKU: string;
    ItemName: string;
    Description: string;
    Quantity: number;
    PricePerUnit: number;
    Discount: number;
    DiscountUnit: string;
    TotalPrice: number;
    RequiredDate: string;
}

export interface Vendor {
    ID: number;
    Name: string;
    Address: string;
    TaxID: string;
    CreatedAt: string;
    UpdatedAt: string;
}

export interface PurchaseOrder {
    ID: number;
    CreatedAt: string;
    UpdatedAt: string;
    DeletedAt: string | null;
    PONumber: string;
    PRID: number;
    VendorID: number;
    Vendor?: Vendor;
    Purpose: string;
    Status: string;
    CreditDay: number;
    DueDate: string;
    IsDeleted: boolean;
    Items: POItem[];
}

export const poApi = {
    getPurchaseOrders: () => {
        
        return request<PurchaseOrder[]>(`/purchase/po`);
    },

    getPurchaseOrderById: (id: number) => request<PurchaseOrder>(`/purchase/po/${id}`),

    createPurchaseOrder: (data: {
        pr_id: number;
        vendor_id: number;
        po_items: Array<{
            pr_item_id: number;
            sku?: string;
            item_name?: string;
            description?: string;
            quantity?: number;
            price_per_unit?: number;
            discount?: number;
            discount_unit?: string;
            required_date?: string;
        }>;
        credit_day?: number;
        due_date: string;
    }) =>
        request<{ message: string; data: PurchaseOrder }>("/purchase/po", {
            method: "POST",
            body: JSON.stringify(data),
        }),

    receiveGoods: (id: number) =>
        request<{ message: string; data: PurchaseOrder }>(`/purchase/po/${id}`, {
            method: "PUT",
        }),

    deletePurchaseOrder: (id: number) =>
        request<{ message: string }>(`/purchase/po/${id}`, {
            method: "DELETE",
        }),
};
